### PR TITLE
Snowbridge Headers on Demand

### PR DESCRIPTION
**Pull request by claravanstaden**
Upgrades Snowbridge to have a much more cost effective Ethereum light client. In the current light client in production, all Ethereum execution headers are stored, whether they contain a transfer message or not. This PR upgrades Snowbridge to not store execution headers at all, but provide the execution header proof alongside a message. Ethereum beacon headers are still stored, to track finalized headers and do ancestry proofs.

More details on the change on the original PR: https://github.com/paritytech/polkadot-sdk/pull/3761

### CODE REVIEW

**Review by bkontur — APPROVED** [review]

**Review by acatangiu — APPROVED** [review]

**Review by acatangiu — APPROVED** [review]

**Comment by claravanstaden** [comment]
@acatangiu the CI wants me to update the AH Kusama spec version: https://github.com/polkadot-fellows/runtimes/actions/runs/9713954965/job/26811939881?pr=365 Not sure if that is OK.

Should the BH spec version be `1_002_008`?

File: system-parachains/asset-hubs/asset-hub-kusama/src/lib.rs [hunk 1]
```diff
@@ -112,7 +112,7 @@ pub const VERSION: RuntimeVersion = RuntimeVersion {
 	spec_name: create_runtime_str!("statemine"),
 	impl_name: create_runtime_str!("statemine"),
 	authoring_version: 1,
-	spec_version: 1_002_006,
+	spec_version: 1_002_008,
```
Comment (acatangiu) [inline system-parachains/asset-hubs/asset-hub-kusama/src/lib.rs:115]: if we massage the dependency bumps a bit we might get away without having to bump the AH spec versions, right @claravanstaden ?

I wouldn't release new AH runtime versions that bring nothing new

**Review by acatangiu — APPROVED** [review]

File: system-parachains/asset-hubs/asset-hub-kusama/src/lib.rs (hunk 1 above)
Comment (acatangiu) [reply system-parachains/asset-hubs/asset-hub-kusama/src/lib.rs:115]: actually no, I'm wrong.

Even if this PR does not change anything for AHs, there have been other things merged in the meantime relevant to AHs.

So all good with bumping this.

**Comment by claravanstaden** [comment]
@acatangiu I removed the P<>K and Snowbridge unstuck migrations, since they ran successfully. Is that OK?

File: system-parachains/asset-hubs/asset-hub-kusama/src/lib.rs (hunk 1 above)
Comment (claravanstaden) [reply system-parachains/asset-hubs/asset-hub-kusama/src/lib.rs:115]: @acatangiu thanks! I see CI is complaining about the Kusama relay version too: https://github.com/polkadot-fellows/runtimes/actions/runs/9740656775/job/26878232348?pr=365

I bumped both the Polkadot and Kusama relay versions: https://github.com/polkadot-fellows/runtimes/pull/365/commits/b2a0b268db849763f1bee19a2ff52a02d0e14fba

Is that OK?

Comment (acatangiu) [reply system-parachains/asset-hubs/asset-hub-kusama/src/lib.rs:115]: yes, all good 👍 we have changelog entries for them too I see

**Comment by claravanstaden** [comment]
@acatangiu what is the process to update the benchmarks? I need to do that before we release. Can I run it manually?

**Comment by bkontur** [comment]
> @acatangiu what is the process to update the benchmarks? I need to do that before we release. Can I run it manually?

@claravanstaden 

Actually, there is no process yet :), but there is some ongoing work on the bot or process https://github.com/polkadot-fellows/runtimes/issues/128 by @mordamax.

You can do it by yourself, here is a tutorial: https://github.com/polkadot-fellows/runtimes/blob/main/docs/weight-generation.md and you need just some machine/VM aligned with reference hardware.

I did the last couple of weights regeneration on my dedicated machine, so I can also do it for you here, just let me know which pallets do you need and I can create PR to this PR.

**Comment by claravanstaden** [comment]
@bkontur thank you for the tips! I would appreciate you helping out since I don't have the required hardware ready off-hand.

I need new weights for these two pallets:
- snowbridge_pallet_ethereum_client
- snowbridge_pallet_inbound_queue

Thanks again! <3

**Comment by bkontur** [comment]
> @bkontur thank you for the tips! I would appreciate you helping out since I don't have the required hardware ready off-hand.
> 
> I need new weights for these two pallets:
> 
>     * snowbridge_pallet_ethereum_client
> 
>     * snowbridge_pallet_inbound_queue
> 
> 
> Thanks again! <3

please, check: https://github.com/Snowfork/runtimes/pull/13

**Review by acatangiu — APPROVED** [review]

**Review by bkchr — APPROVED** [review]

**Pull request merged**